File: src/processing/state_validator.py
"""
State validation for reset operations.
Ensures state integrity after each operation.
"""

import numpy as np
from typing import Dict, Any, Optional
import logging
from .reset_transaction import ResetOperation

logger = logging.getLogger(__name__)


class StateValidator:
    """
    Validates state integrity after reset operations.

    Performs checks for:
    - Required fields presence
    - Data type correctness
    - Value range validity
    - NaN/Inf detection
    - Structural consistency
    """

    def validate(self, state: Dict[str, Any], operation: ResetOperation) -> bool:
        """
        Validate state based on operation type.

        Args:
            state: State to validate
            operation: Type of operation performed

        Returns:
            True if state is valid, False otherwise
        """
        validators = {
            ResetOperation.KALMAN_RESET: self._validate_kalman_state,
            ResetOperation.STATE_UPDATE: self._validate_processor_state,
            ResetOperation.BUFFER_UPDATE: self._validate_buffer_state,
            ResetOperation.STATE_PERSIST: self._validate_persisted_state
        }

        validator = validators.get(operation)
        if not validator:
            logger.error(f"No validator for operation {operation}")
            return False

        try:
            return validator(state)
        except Exception as e:
            logger.error(f"Validation failed for {operation}: {e}")
            return False

    def _validate_kalman_state(self, state: Dict) -> bool:
        """
        Validate Kalman filter state after reset.

        Checks:
        - State has been properly reset (kalman_params should be None after reset)
        - Reset parameters are present and valid
        - Measurements counter is reset
        """
        # After a reset, kalman_params should be None (will be recreated on next measurement)
        if state.get('kalman_params') is not None:
            logger.warning("kalman_params should be None after reset")
            # This is actually OK - might be set during processing

        # Check reset parameters
        if 'reset_parameters' not in state:
            logger.error("Missing reset_parameters after reset")
            return False

        reset_params = state['reset_parameters']
        required_reset_params = [
            'initial_variance_multiplier', 'weight_noise_multiplier',
            'trend_noise_multiplier', 'observation_noise_multiplier',
            'adaptation_measurements', 'adaptation_days', 'adaptation_decay_rate',
            'quality_acceptance_threshold'
        ]

        for param in required_reset_params:
            if param not in reset_params:
                logger.error(f"Missing reset parameter: {param}")
                return False

            value = reset_params[param]

            # Check for NaN or Inf in numeric parameters
            if isinstance(value, (int, float)):
                if np.isnan(value) or np.isinf(value):
                    logger.error(f"Reset parameter {param} is NaN or Inf")
                    return False

                # Range checks for multipliers
                if 'multiplier' in param:
                    if value <= 0:
                        logger.error(f"Reset parameter {param} must be positive, got {value}")
                        return False

                if param == 'adaptation_decay_rate':
                    if value <= 0:
                        logger.error(f"Adaptation decay rate must be positive: {value}")
                        return False

        # Check measurements counter
        if state.get('measurements_since_reset', -1) != 0:
            logger.error(f"Measurements counter should be 0 after reset, got {state.get('measurements_since_reset')}")
            return False

        # Check reset type is valid (handle both upper and lowercase)
        reset_type = state.get('reset_type')
        valid_types = ['INITIAL', 'HARD', 'SOFT', 'initial', 'hard', 'soft', None]
        if reset_type not in valid_types:
            logger.error(f"Invalid reset type: {reset_type}")
            return False

        # Check reset timestamp exists
        if 'reset_timestamp' not in state:
            logger.error("Missing reset_timestamp")
            return False

        return True

    def _validate_processor_state(self, state: Dict) -> bool:
        """
        Validate overall processor state update.

        Checks for consistency and required fields.
        """
        # Essential fields that should always exist
        essential_fields = [
            'measurements_since_reset',
            'reset_type',
            'reset_parameters',
            'reset_timestamp'
        ]

        for field in essential_fields:
            if field not in state:
                logger.error(f"Missing essential field: {field}")
                return False

        # Validate measurements counter
        measurements = state.get('measurements_since_reset', -1)
        if measurements < 0:
            logger.error(f"Invalid measurements count: {measurements}")
            return False

        # Validate measurement history if present
        if 'measurement_history' in state:
            history = state['measurement_history']
            if not isinstance(history, list):
                logger.error("Measurement history is not a list")
                return False

            # Check history isn't too large (memory protection)
            max_history_size = 1000
            if len(history) > max_history_size:
                logger.error(f"Measurement history too large: {len(history)} > {max_history_size}")
                return False

        # If there's a last_state, validate it's numeric
        if 'last_state' in state and state['last_state'] is not None:
            last_state = state['last_state']
            if not isinstance(last_state, (int, float, np.ndarray)):
                logger.error(f"Invalid last_state type: {type(last_state)}")
                return False

            # Check for NaN/Inf
            if isinstance(last_state, (int, float)):
                if np.isnan(last_state) or np.isinf(last_state):
                    logger.error("last_state contains NaN or Inf")
                    return False

        return True

    def _validate_buffer_state(self, state: Dict) -> bool:
        """
        Validate buffer state after update.

        Checks buffer consistency and size limits.
        """
        # Buffer operations are handled differently in this system
        # After a reset, measurement_history should be cleared
        if 'measurement_history' in state:
            history = state.get('measurement_history', [])

            # After reset, history should be empty or very small
            if len(history) > 100:  # Reasonable limit after reset
                logger.warning(f"Large measurement history after reset: {len(history)} items")
                # Not a failure, just a warning

        return True

    def _validate_persisted_state(self, state: Dict) -> bool:
        """
        Validate state before persistence.

        Final validation before saving to database.
        """
        # Run all validations for a complete check

        # Check structure
        if not isinstance(state, dict):
            logger.error("State is not a dictionary")
            return False

        # Check for any NaN/Inf in numeric fields
        for key, value in state.items():
            if isinstance(value, (int, float)):
                if np.isnan(value) or np.isinf(value):
                    logger.error(f"Field {key} contains NaN or Inf")
                    return False

            elif isinstance(value, np.ndarray):
                if np.any(np.isnan(value)) or np.any(np.isinf(value)):
                    logger.error(f"Array field {key} contains NaN or Inf")
                    return False

        return True

    @staticmethod
    def validate_weight_value(weight: float) -> bool:
        """
        Validate a weight measurement value.

        Args:
            weight: Weight value in kg

        Returns:
            True if valid weight
        """
        # Basic sanity checks
        if np.isnan(weight) or np.isinf(weight):
            logger.error("Weight is NaN or Inf")
            return False

        # Physiological limits (same as in constants)
        if not (20 <= weight <= 450):
            logger.error(f"Weight {weight}kg outside physiological limits")
            return False

        return True

    @staticmethod
    def validate_reset_type(reset_type: str) -> bool:
        """
        Validate reset type string.

        Args:
            reset_type: Reset type string

        Returns:
            True if valid reset type
        """
        valid_types = ['INITIAL', 'HARD', 'SOFT', 'initial', 'hard', 'soft']
        return reset_type in valid_types

File: typescript_lib/src/weight-processor-lib/core/processing/state_validator.ts
/**
 * State validation for reset operations.
 * Ensures state integrity after each operation.
 */

/**
 * Types of operations in a reset transaction
 */
export enum ResetOperation {
  KALMAN_RESET = 'kalman_reset',
  STATE_UPDATE = 'state_update',
  BUFFER_UPDATE = 'buffer_update',
  STATE_PERSIST = 'state_persist',
}

/**
 * Type for processor state dictionary
 */
export interface ProcessorState {
  kalman_params?: any;
  reset_parameters?: ResetParameters;
  measurements_since_reset?: number;
  reset_type?: string;
  reset_timestamp?: number;
  measurement_history?: any[];
  last_state?: number | number[];
  [key: string]: any;
}

/**
 * Type for reset parameters
 */
export interface ResetParameters {
  initial_variance_multiplier: number;
  weight_noise_multiplier: number;
  trend_noise_multiplier: number;
  observation_noise_multiplier: number;
  adaptation_measurements: number;
  adaptation_days: number;
  adaptation_decay_rate: number;
  quality_acceptance_threshold: number;
  [key: string]: any;
}

/**
 * Check if a value is NaN or Infinity
 */
function isNaNOrInf(value: number): boolean {
  return isNaN(value) || !isFinite(value);
}

/**
 * Check if an array contains NaN or Infinity
 */
function arrayHasNaNOrInf(arr: number[]): boolean {
  return arr.some(val => isNaNOrInf(val));
}

/**
 * Validates state integrity after reset operations.
 *
 * Performs checks for:
 * - Required fields presence
 * - Data type correctness
 * - Value range validity
 * - NaN/Inf detection
 * - Structural consistency
 */
export class StateValidator {
  /**
   * Validate state based on operation type.
   *
   * @param state - State to validate
   * @param operation - Type of operation performed
   * @returns True if state is valid, False otherwise
   */
  validate(state: ProcessorState, operation: ResetOperation): boolean {
    const validators: Record<ResetOperation, (state: ProcessorState) => boolean> = {
      [ResetOperation.KALMAN_RESET]: this._validateKalmanState.bind(this),
      [ResetOperation.STATE_UPDATE]: this._validateProcessorState.bind(this),
      [ResetOperation.BUFFER_UPDATE]: this._validateBufferState.bind(this),
      [ResetOperation.STATE_PERSIST]: this._validatePersistedState.bind(this),
    };

    const validator = validators[operation];
    if (!validator) {
      console.error(`No validator for operation ${operation}`);
      return false;
    }

    try {
      return validator(state);
    } catch (e) {
      console.error(`Validation failed for ${operation}:`, e);
      return false;
    }
  }

  /**
   * Validate Kalman filter state after reset.
   *
   * Checks:
   * - State has been properly reset (kalman_params should be None after reset)
   * - Reset parameters are present and valid
   * - Measurements counter is reset
   */
  private _validateKalmanState(state: ProcessorState): boolean {
    // After a reset, kalman_params should be None (will be recreated on next measurement)
    if (state.kalman_params !== undefined && state.kalman_params !== null) {
      console.warn('kalman_params should be null after reset');
      // This is actually OK - might be set during processing
    }

    // Check reset parameters
    if (!state.reset_parameters) {
      console.error('Missing reset_parameters after reset');
      return false;
    }

    const resetParams = state.reset_parameters;
    const requiredResetParams = [
      'initial_variance_multiplier',
      'weight_noise_multiplier',
      'trend_noise_multiplier',
      'observation_noise_multiplier',
      'adaptation_measurements',
      'adaptation_days',
      'adaptation_decay_rate',
      'quality_acceptance_threshold',
    ];

    for (const param of requiredResetParams) {
      if (!(param in resetParams)) {
        console.error(`Missing reset parameter: ${param}`);
        return false;
      }

      const value = resetParams[param];

      // Check for NaN or Inf in numeric parameters
      if (typeof value === 'number') {
        if (isNaNOrInf(value)) {
          console.error(`Reset parameter ${param} is NaN or Inf`);
          return false;
        }

        // Range checks for multipliers
        if (param.includes('multiplier')) {
          if (value <= 0) {
            console.error(`Reset parameter ${param} must be positive, got ${value}`);
            return false;
          }
        }

        if (param === 'adaptation_decay_rate') {
          if (value <= 0) {
            console.error(`Adaptation decay rate must be positive: ${value}`);
            return false;
          }
        }
      }
    }

    // Check measurements counter
    if ((state.measurements_since_reset ?? -1) !== 0) {
      console.error(
        `Measurements counter should be 0 after reset, got ${state.measurements_since_reset}`
      );
      return false;
    }

    // Check reset type is valid (handle both upper and lowercase)
    const resetType = state.reset_type;
    const validTypes = ['INITIAL', 'HARD', 'SOFT', 'initial', 'hard', 'soft', null, undefined];
    if (!validTypes.includes(resetType as any)) {
      console.error(`Invalid reset type: ${resetType}`);
      return false;
    }

    // Check reset timestamp exists
    if (!('reset_timestamp' in state)) {
      console.error('Missing reset_timestamp');
      return false;
    }

    return true;
  }

  /**
   * Validate overall processor state update.
   *
   * Checks for consistency and required fields.
   */
  private _validateProcessorState(state: ProcessorState): boolean {
    // Essential fields that should always exist
    const essentialFields = [
      'measurements_since_reset',
      'reset_type',
      'reset_parameters',
      'reset_timestamp',
    ];

    for (const field of essentialFields) {
      if (!(field in state)) {
        console.error(`Missing essential field: ${field}`);
        return false;
      }
    }

    // Validate measurements counter
    const measurements = state.measurements_since_reset ?? -1;
    if (measurements < 0) {
      console.error(`Invalid measurements count: ${measurements}`);
      return false;
    }

    // Validate measurement history if present
    if ('measurement_history' in state) {
      const history = state.measurement_history;
      if (!Array.isArray(history)) {
        console.error('Measurement history is not an array');
        return false;
      }

      // Check history isn't too large (memory protection)
      const maxHistorySize = 1000;
      if (history.length > maxHistorySize) {
        console.error(
          `Measurement history too large: ${history.length} > ${maxHistorySize}`
        );
        return false;
      }
    }

    // If there's a last_state, validate it's numeric
    if ('last_state' in state && state.last_state !== null && state.last_state !== undefined) {
      const lastState = state.last_state;
      if (typeof lastState === 'number') {
        // Check for NaN/Inf
        if (isNaNOrInf(lastState)) {
          console.error('last_state contains NaN or Inf');
          return false;
        }
      } else if (Array.isArray(lastState)) {
        // Check array for NaN/Inf
        if (arrayHasNaNOrInf(lastState)) {
          console.error('last_state array contains NaN or Inf');
          return false;
        }
      } else {
        console.error(`Invalid last_state type: ${typeof lastState}`);
        return false;
      }
    }

    return true;
  }

  /**
   * Validate buffer state after update.
   *
   * Checks buffer consistency and size limits.
   */
  private _validateBufferState(state: ProcessorState): boolean {
    // Buffer operations are handled differently in this system
    // After a reset, measurement_history should be cleared
    if ('measurement_history' in state) {
      const history = state.measurement_history ?? [];

      // After reset, history should be empty or very small
      if (history.length > 100) {
        // Reasonable limit after reset
        console.warn(
          `Large measurement history after reset: ${history.length} items`
        );
        // Not a failure, just a warning
      }
    }

    return true;
  }

  /**
   * Validate state before persistence.
   *
   * Final validation before saving to database.
   */
  private _validatePersistedState(state: ProcessorState): boolean {
    // Run all validations for a complete check

    // Check structure
    if (typeof state !== 'object' || state === null || Array.isArray(state)) {
      console.error('State is not an object');
      return false;
    }

    // Check for any NaN/Inf in numeric fields
    for (const [key, value] of Object.entries(state)) {
      if (typeof value === 'number') {
        if (isNaNOrInf(value)) {
          console.error(`Field ${key} contains NaN or Inf`);
          return false;
        }
      } else if (Array.isArray(value)) {
        // Check if it's a numeric array
        const isNumericArray = value.every(v => typeof v === 'number');
        if (isNumericArray && arrayHasNaNOrInf(value)) {
          console.error(`Array field ${key} contains NaN or Inf`);
          return false;
        }
      }
    }

    return true;
  }

  /**
   * Validate a weight measurement value.
   *
   * @param weight - Weight value in kg
   * @returns True if valid weight
   */
  static validateWeightValue(weight: number): boolean {
    // Basic sanity checks
    if (isNaNOrInf(weight)) {
      console.error('Weight is NaN or Inf');
      return false;
    }

    // Physiological limits (same as in constants)
    if (weight < 20 || weight > 450) {
      console.error(`Weight ${weight}kg outside physiological limits`);
      return false;
    }

    return true;
  }

  /**
   * Validate reset type string.
   *
   * @param resetType - Reset type string
   * @returns True if valid reset type
   */
  static validateResetType(resetType: string): boolean {
    const validTypes = ['INITIAL', 'HARD', 'SOFT', 'initial', 'hard', 'soft'];
    return validTypes.includes(resetType);
  }
}
